SurveyAnswerChart: Add support for a dynamic interval type.

greinard opened the pull request:
## Overview

> Explain your changes, including any issues or relevant context about why they are needed.

This branch adds support for a new "Dynamic" interval type to the `SurveyAnswerChart`.

<img width="595" height="250" alt="image" src="https://github.com/user-attachments/assets/f6bf6162-2bd4-43b8-97cd-3986fae41ec5" />

When using the dynamic interval type, the chart will query for all survey answers (no time filter) and will establish the x-axis domain using the actual data rather than using a fixed time range.

There is also a secondary setting, the `dynamicIntervalEndType` that can be set to `Today` or `Last Reading` to adjust the end date of the x-axis domain.  When using `Today`, the x-axis domain will range from the earliest survey answer date through today.  When using `Last Reading`, the x-axis domain will range from the earliest survey answer date through the most recent survey answer date.  The default is `Today` when no end type has been specified.

I'll revert the snapshot version prior to merging.

## Security

> Consider potential security impacts and complete the following checklist. 
> **REMINDER:** All file contents are public.

- [x] I have ensured no secure credentials or sensitive information remain in code, metadata, comments, etc.
  - [x] Please verify that you double checked that .storybook/preview.js does not contain your participant access key details. 
  - [x] There are no temporary testing changes committed such as API base URLs, access tokens, print/log statements, etc.
- [x] These changes do not introduce any security risks, or any such risks have been properly mitigated.

DESCRIBE BRIEFLY WHAT SECURITY RISKS YOU CONSIDERED, WHY THEY DON'T APPLY, OR HOW THEY'VE BEEN MITIGATED.

- No new security risk.  This change is an optional filtering and rendering adjustment for data that is already available via the exiting components and APIs.

## Testing

> Consider whether the changes might have device-specific behaviors (screen padding, new APIs, etc.) and check one of the following boxes:

- [ ] This change can be adequately tested using the MDH Storybook.
- [x] This change requires additional testing in the MDH iOS/Android/Web apps. (Create a pre-release tag/build and test in a ViewBuilder PR.)

The majority of this change can be adequately tested using the Storybook, but you'll need to run it in live mode or within a VB view to test the non-preview querying updates.

### Documentation

> Consider whether there are any documentation impacts and check one of the following boxes:

- [x] I have added relevant Storybook updates to this PR.
- [ ] If this feature requires a developer doc update, I have tagged `@CareEvolution/api-docs`.
- [ ] This change does not impact documentation or Storybook.

felideon commented:
Tested the new Dynamic interval through ViewBuilder with a test participant I had for the BD² project, that has some scores that we compute (PHQ-9, PMQ-9, etc.). I'm noticing a minor issue when using an area chart and multiple series with a mix of data gaps that doesn't show up in the Storybook:

<img width="357" height="445" alt="Screenshot 2025-08-28 at 4 24 00 PM" src="https://github.com/user-attachments/assets/9a0fbd81-dc22-4e21-97ba-59160020d339" />

Note that in the top area chart there is no apparent data point at the beginning start of the series for the who5 series (as it's the only data point for that series). You can see it on hover/tap, though:
<img width="352" height="447" alt="Screenshot 2025-08-28 at 4 25 28 PM" src="https://github.com/user-attachments/assets/73d48c46-75e1-40cf-9d21-80e6b0a55b0b" />

I wasn't sure what it should look like, but the "with one data point" story does show dots in that case. Let me know if you want the test participant info for testing/debugging.

Other than that, everything looks good and the code changes are sound to me. The new preview data generator is pretty nice.

felideon commented:
That's unfortunate since I think the area charts look nicer without the dots, specially on busier ones. Given it's a Recharts limitation, I guess we'll have to live with it. 

cc @ajmenca @chrisnowak for awareness

felideon commented:
+1 for the new UI stories, will check in View Builder again once it's pulled in on that [branch](https://github.com/CareEvolution/MyDataHelpsViewBuilder/pull/354).

chrisnowak commented on src/components/container/SurveyAnswerChart/SurveyAnswerChart.tsx:
```diff
@@ -24,180 +21,139 @@ export interface SurveyAnswerAreaChartSeries extends AreaChartSeries {
-            });
-            return;
+
+    const intervalType = useMemo((): 'Week' | 'Month' | '6Month' | 'Dynamic' => {
```
(All of these seem pretty cheap to calculate and so don't seem worth the added complexity)

greinard replied:
This may be a tad bit long winded, but here's how I arrived at that.  🙂 

The `initialize` function, formerly named `loadCurrentInterval`, depends upon the computed `intervalStart` and `intervalEnd` values to do what it needs to do along with a couple other dependencies.

However, the dependencies used for triggering calls to the `loadCurrentInterval` function did not include the computed `intervalStart` and `intervalEnd` values, but rather, included the things that went into computing those values.  This felt a little disconnected to me, since those dependencies were not actually referenced in `loadCurrentInterval` at all.  It was just a loose coupling because someone happened to know that those other dependencies were part of computing the `intervalStart` and `intervalEnd` values which are actually being used within the `loadCurrentInterval` function.

So in an attempt to make things more clear with regard to the hierarchy of dependencies, I adjusted the dependencies for `initialize` to use the computed `intervalStart` and `intervalEnd` instead.  However, each time those dates get computed, even if they represent the same date, they fail the reference equality check on the dependency list, so they cause `initialize` to be called over and over.  The way to prevent that is with `useMemo`, which caches and returns the same computed date until it actually needs to be updated because one of *its* dependencies changed.

So `useMemo` was only actually needed for `intervalStart` and `intervalEnd` to make this work, but I just kept things consistent and used it for `intervalType` here as well.

We can drop the use of `useMemo` on these, just turning them into standard IIFEs, but then I'll need to add their dependencies back to the dependencies list for the `initialize` call.

So I guess the complexity can go in either direction.  We can add the use of another hook that needs to be understood, but clearly defines the dependency hierarchy and avoids unnecessary re-computation, or we can use standard IIFEs for the computations (since they are very simple) and just have a few loosely coupled dependencies in the list of dependencies for the `initialize` function.

Thoughts?

greinard commented:
I fixed the issue with the dynamic bar chart (using last reading) rendering two data points.  Apparently recharts has trouble rendering the bar charts with two data points when they are exactly at the domain boundary.  I fixed it by adding a small buffer to the x-axis domain in that specific situation, similar to how we handle/center a single data point.

greinard commented:
@chrisnowak - I'm going to proceed with this as-is for now, but if you feel strongly that we should drop the `useMemo` hooks, please let me know and I can do so in a follow up PR.  I prefer the `useMemo` impl, but I don't feel that strongly about it since it'll work either way.
